src/os/bluestore/BlueStore.cc

```diff
@@ -1810,37 +1858,37 @@ void BlueStore::BufferSpace::read(
   cache->logger->inc(l_bluestore_buffer_miss_bytes, miss_bytes);
 }
 
-void BlueStore::BufferSpace::_finish_write(BufferCacheShard* cache,  uint64_t seq)
+void BlueStore::BufferSpace::_finish_write(BufferCacheShard* cache,
+                                           uint64_t seq,
+                                           uint32_t offset, uint32_t len)
 {
-  ldout(cache->cct, 20) << __func__ << " seq=" << seq << dendl;
-  auto i = writing.begin();
-  while (i != writing.end()) {
-    if (i->seq > seq) {
-      break;
-    }
-    if (i->seq < seq) {
-      ++i;
-      continue;
-    }
-
-    Buffer *b = &*i;
-    ceph_assert(b->is_writing());
+  ldout(cache->cct, 10) << __func__ << " seq " << seq
+                        << std::hex << " 0x" << offset << "~" << len << std::dec
+                        << dendl;
 
-    if (b->flags & Buffer::FLAG_NOCACHE) {
-      writing.erase(i++);
-      ldout(cache->cct, 20) << __func__ << " discard " << *b << dendl;
-      buffer_map.erase(b->offset);
-    } else {
-      b->state = Buffer::STATE_CLEAN;
-      writing.erase(i++);
-      b->maybe_rebuild();
-      b->data.reassign_to_mempool(mempool::mempool_bluestore_cache_data);
-      cache->_add(b, 1, nullptr);
-      ldout(cache->cct, 20) << __func__ << " added " << *b << dendl;
+  uint32_t end = offset + len;
+  std::lock_guard l(cache->lock);
+  auto i = _data_lower_bound(offset);
+  while (i != buffer_map.end() && offset < end && i->offset < end) {
+    Buffer* b = &*i;
+    i++;
+    ceph_assert(b->end() > offset);
+    if (b->seq == seq && b->is_writing()) {
```

Comment:
We used to have some complicated processing scheme in `_finish_write` that used `i->seq<seq` and `i->seq>seq` etc.
Now it is simplified to `i->seq == seq` and we process just those and ignore rest. I like it
But if it is so, we could change
`uint64_t Buffer::seq` into `TransContext* Buffer::ctx`.

The deal is that we could now call `finish_write(ctx)`;
each onode in ctx's modified_onodes will inform its BufferSpace;
each buffers in onodes's BufferSpace from that ctx will move from `STATE_WRITING` to `STATE_CLEAN`.
No change here.
The difference is the reaction to `clone()` with buffers in `STATE_WRITING`.
We could just copy buffers from onode src to onode dst, and then update TransContext::modified_onodes.

@ifed01 What do you think?

Reply:
I like this.

Reply:
LGTM too.